src/wallet/bdb.cpp

```diff
@@ -315,11 +315,17 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const bool read_only, b
     env = database.env.get();
     pdb = database.m_db.get();
     strFile = fs::PathToString(database.m_filename);
-    if (!Exists(std::string("version"))) {
-        bool fTmp = fReadOnly;
-        fReadOnly = false;
-        Write(std::string("version"), CLIENT_VERSION);
-        fReadOnly = fTmp;
+
+    // At the first db open
+    if (database.client_version == -1) {
```

Comment:
Mainly because of how the flow is currently structured. We are actually checking if the db is open on every new batch, and opening it if not (that is done few lines above, at line 312).

In other words and to be a bit more specific: I was going to add it inside the `BerkeleyDatabase::Open()` method but.. as `BerkeleyBatch` contains the functions to access the db (Read/Write/Erase etc), I was going to end up duplicating the code there (the `BerkeleyBatch::ReadKey` functionality).

Not sure if I should include it here but I think that a good next step would be to decouple the db.open call from the `BerkeleyBatch` constructor, so it only receives an already opened database. And in that way, the client version read could be decoupled as well and not be part of the constructor.

Reply:
Fair enough. If you have a good reason for it it's okay with me. I was just asking.
It's "only" the legacy wallet so doing any large refactors is probably not worth it.

Reply:
yeah, no problem. The question was great to be sync about following-up improvements. The SQLite flow has some similarities.. (the db constructor is, mostly, a no-op as the db is always opened etc..).

plus, I'm grateful for your quick feedback :).